evolver/app/tests/test_app.py

```diff
@@ -25,9 +29,25 @@ def test_evolver_app_default_config_dump_endpoint(self, app_client):
         assert response.status_code == 200
         assert sorted(response.json().keys()) == ['config', 'last_read', 'state']
 
+    def test_EvolverConfigWithoutDefaults(self):
+        EvolverConfigWithoutDefaults.model_validate(EvolverConfig().model_dump())
+        EvolverConfigWithoutDefaults.model_validate_json(EvolverConfig().model_dump_json())
+
     def test_evolver_update_config_endpoint(self, app_client):
         data = {'hardware': {'test': {'driver': 'evolver.hardware.demo.NoOpSensorDriver'}}}
         response = app_client.post('/', json=data)
+        # all config fields are required so the above post failed with an Unprocessable Entity error.
+        assert response.status_code == 422
+        contents = json.loads(response.content)
+        for content in contents["detail"]:
+            assert content["msg"] == "Field required"
+
+        new_data = EvolverConfig().copy(update=data)
+        # Note: pydantic.BaseModel.model_dump() doesn't serialize the driver class and canonical json has issues
+        # serializing some classes, e.g., ABCMeta. However, pydantic.BaseModel.model_dump_json() knows how to convert
+        # pydantic.ImportString. However, some setting must be adrift as this results in a 422 "Input should be a valid
+        # dictionary or object to extract fields from". To resolve this, we convert it back to a dict.
+        response = app_client.post('/', json=json.loads(new_data.model_dump_json()))
```

Comment:
``app_client.post('/', json=json.dumps(json.loads(new_data.model_dump_json())))`` also gives a 422 so looks like a ``post`` issue.

Reply:
Yep, https://github.com/tiangolo/fastapi/issues/192.

The docs are very misleading, especially since [the docs](https://fastapi.tiangolo.com/reference/testclient/?h=testclient#fastapi.testclient.TestClient.post) state that ``json`` can be of any type. Now, I'm getting Deja Vu and remember hitting this issue in the past.

Reply:
I think this matches the call for requests, where the json field is processed into json (from a python object), so you could probably achieve the same thing as here (without multiple serializations) with:

```py
 response = app_client.post('/', data=new_data.model_dump_json())
```